### PR TITLE
[dv/common] TLUL self-check testbench

### PR DESCRIPTION
**Pull request by weicaiyang**
Create a TB in order to verify some features which isn't supported by
design, like device same cycle response

This TB contains a scoreboard to check if req/rsp are correct
Can be used for reference to develop the other agent TB

Here is the waveform with +zero_delays=1 and same cycle response enabled
Screen Shot 2020-09-17 at 4 08 28 PM

Signed-off-by: Weicai Yang <weicai@google.com>

### CODE REVIEW

**Comment by tunghoang290780** [comment]
@weicaiyang can we set zero_delays to 0? Otherwise, LGTM

File: hw/dv/sv/tl_agent/dv/env/seq_lib/tl_agent_base_vseq.sv [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+// ---------------------------------------------
+// tl_agent environment virtual sequence
+// ---------------------------------------------
+class tl_agent_base_vseq extends dv_base_vseq #(.CFG_T               (tl_agent_env_cfg),
+                                                .COV_T               (tl_agent_env_cov),
+                                                .VIRTUAL_SEQUENCER_T (tl_agent_virtual_sequencer));
+  uint min_req_cnt = 100;
+  uint max_req_cnt = 200;
```
Comment (tunghoang290780) [inline hw/dv/sv/tl_agent/dv/env/seq_lib/tl_agent_base_vseq.sv:12]: could we define these variables in cfg?

File: hw/dv/sv/tl_agent/dv/env/seq_lib/tl_agent_base_vseq.sv [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    host_seq = tl_host_seq::type_id::create("host_seq");
+    `DV_CHECK_RANDOMIZE_WITH_FATAL(host_seq,
+                                   req_cnt inside {[min_req_cnt : max_req_cnt]};)
+    host_seq.start(p_sequencer.host_seqr);
+    `uvm_info(`gfn, $sformatf("host_seq finished sending %0d requests", host_seq.req_cnt), UVM_LOW)
+  endtask
+
+  virtual task body();
+    run_device_seq_nonblocking();
+    for (int i = 1; i <= num_trans; i++) begin
```
Comment (rswarbrick) [inline hw/dv/sv/tl_agent/dv/env/seq_lib/tl_agent_base_vseq.sv:40]: Not that it really matters, but 1-based `for` loops are rather unusual. I'd probably start at 0 (or use `repeat`).

File: hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv [hunk 3]
```diff
@@ -0,0 +1,71 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+// ---------------------------------------------
+// tl_agent environment class
+// ---------------------------------------------
+class tl_agent_env extends dv_base_env#(.CFG_T              (tl_agent_env_cfg),
```
Comment (rswarbrick) [inline hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv:8]: Nit: Missing space before `#` (see [style guide](https://github.com/lowRISC/style-guides/blob/master/VerilogCodingStyle.md#parameterized-types)).

Also, I'd personally add one more space before each `(` in the table: since it's tabulated, the `.VIRTUAL_SEQUENCER_T` line looks a little odd to me. No worries if you disagree, though.

File: hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                                        .COV_T              (tl_agent_env_cov));
+
+  tl_agent host_agent;
+  tl_agent device_agent;
+
+  `uvm_component_utils(tl_agent_env)
+
+  function new (string name, uvm_component parent);
+    super.new(name, parent);
+  endfunction : new
```
Comment (rswarbrick) [inline hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv:20]: I think this could be done with the `uvm_component_new` macro.

File: hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    if (cfg.zero_delays) begin
+      cfg.host_agent_cfg.a_valid_delay_min = 0;
+      cfg.host_agent_cfg.a_valid_delay_max = 0;
+      cfg.host_agent_cfg.d_ready_delay_min = 0;
+      cfg.host_agent_cfg.d_ready_delay_max = 0;
+      cfg.device_agent_cfg.d_valid_delay_min = 0;
+      cfg.device_agent_cfg.d_valid_delay_max = 0;
+      cfg.device_agent_cfg.a_ready_delay_min = 0;
+      cfg.device_agent_cfg.a_ready_delay_max = 0;
+    end
```
Comment (rswarbrick) [inline hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv:40]: Shouldn't this logic be in the configuration's `initialize` function, or maybe as a constraint that's applied when randomizing the configuration?

File: hw/dv/sv/tl_agent/dv/tests/tl_agent_base_test.sv [hunk 6]
```diff
@@ -0,0 +1,15 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+class tl_agent_base_test extends dv_base_test #(.ENV_T(tl_agent_env), .CFG_T(tl_agent_env_cfg));
+  `uvm_component_utils(tl_agent_base_test)
+  `uvm_component_new
+
+  virtual function void build_phase(uvm_phase phase);
+    max_quit_count  = 50;
```
Comment (rswarbrick) [inline hw/dv/sv/tl_agent/dv/tests/tl_agent_base_test.sv:10]: Why is this required? If I understand correctly, this means we'll stumble on past the first error. It can be set with a plusarg for debugging, but I'm not convinced it helps by default. Have I missed something?

File: hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv (hunk 5 above)
Comment (weicaiyang) [reply hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv:37]: `cfg.zero_delays` can be set from plusargs.
we have below steps to configure these parameters. we finalize the zero_delays in test:build_phase and finalize delays in env:build_phase

```
test:build_phase
  cfg.initialize(); // create RAL, set base addr, size
  cfg.randomize(); // randomize cfg.zero_delay
  void'($value$plusargs("zero_delays=%0b", cfg.zero_delays)); // override from plusargs

env: build_phase
  use final cfg parameters to set the other variables.
```

File: hw/dv/sv/tl_agent/dv/tl_agent_sim.core [hunk 7]
```diff
@@ -0,0 +1,24 @@
+CAPI=2:
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+name: "lowrisc:dv:tl_agent_sim:0.1"
+description: "tl_agent DV sim target"
+filesets:
+  files_rtl:
+    depend:
```
Comment (cindychip) [inline hw/dv/sv/tl_agent/dv/tl_agent_sim.core:9]: Looks like this is empty, do we still need this? and line 22.

File: hw/dv/sv/tl_agent/dv/tests/tl_agent_base_test.sv (hunk 6 above)
Comment (weicaiyang) [reply hw/dv/sv/tl_agent/dv/tests/tl_agent_base_test.sv]: personal preference to have sim run longer rather than stop at the first error. removed this as no needed in this simple TB.

File: hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv (hunk 3 above)
Comment (weicaiyang) [reply hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv]: Fixed, thanks

**Review by cindychip — APPROVED** [review]

File: hw/dv/sv/tl_agent/dv/env/seq_lib/tl_agent_base_vseq.sv (hunk 1 above)
Comment (tunghoang290780) [reply hw/dv/sv/tl_agent/dv/env/seq_lib/tl_agent_base_vseq.sv:12]: LGTM

**Comment by tunghoang290780** [comment]
@weicaiyang LGTM, thanks for your efforts

**Review by tunghoang290780 — APPROVED** [review]

File: hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv (hunk 5 above)
Comment (rswarbrick) [reply hw/dv/sv/tl_agent/dv/env/tl_agent_env.sv:37]: Oh, I see. Thanks for explaining.

**Review by rswarbrick — APPROVED** [review]

File: hw/dv/sv/tl_agent/dv/env/seq_lib/tl_agent_base_vseq.sv (hunk 2 above)
Comment (rasmus-madsen) [reply hw/dv/sv/tl_agent/dv/env/seq_lib/tl_agent_base_vseq.sv]: @weicaiyang I agree with @rswarbrick that this is rather unusual.
I am ok with leaving as is.
but I would probably stop and think if I read in the log that the iteration started with 1 vs 0...

**Review by rasmus-madsen — APPROVED** [review]
LGTM

File: hw/dv/sv/tl_agent/dv/env/tl_agent.core [hunk 8]
```diff
@@ -0,0 +1,26 @@
+CAPI=2:
```
Comment (sriyerg) [inline hw/dv/sv/tl_agent/dv/env/tl_agent.core:1]: Perhaps rename this file to `tl_agent_env.core` to avoid confusion with `tl_agent/tl_agent.core`

**Review by sriyerg — APPROVED** [review]
LGTM save for 2 nits.
